### PR TITLE
Record nodes access optimization

### PR DESCRIPTION
**Pull request by fefrain**

### CODE REVIEW

File: common/record/_internal/recordNodesIndex.js [hunk 1]
```diff
@@ -74,60 +91,84 @@ const addNodes = nodes =>
...
       R.values,
       R.reduce(
-        (acc, node) => Node.isDeleted(node)
-          ? acc
-          : _addNode(node)(acc),
+        (recordAcc, node) =>
+          R.ifElse(
+            R.always(Node.isDeleted(node)),
+            R.identity,
+            _addNode(node)
```
Comment (minotogna) [inline common/record/_internal/recordNodesIndex.js]: Can this just be ```Node.isDeleted(node)? _addNode(node) : recordAcc```

File: common/record/_internal/recordNodesIndex.js [hunk 2]
```diff
@@ -74,60 +91,84 @@ const addNodes = nodes =>
...
+    R.always(add),
+    R.ifElse(
+      R.includes(value),
+      R.identity,
+      R.append(value)
+    ),
+    R.without(value)
+  ),
+  arr => R.assocPath(path, arr, obj)
+)(obj)
```
Comment (minotogna) [inline common/record/_internal/recordNodesIndex.js]: It's not very clear to me have a method that does both. Can we have two? Like ```_assocToIndexPath``` and ```_dissocToIndexPath```?

File: common/record/_internal/recordReader.js [hunk 3]
```diff
@@ -141,55 +125,32 @@ const getDependentNodes = (survey, node, dependencyType) => record => {
...
-      }
-    )
+    for (const ancestor of ancestors) {
+      const children = getNodeChildren(ancestor)(record)
+      const nodeFound = children.find(node => Node.getNodeDefUuid(node) === parentCodeDefUuid)
+      if (nodeFound)
+        return nodeFound
+    }
   }
-
```
Comment (minotogna) [inline common/record/_internal/recordReader.js]: Why has this changed?

File: common/record/_internal/recordUpdater.js [hunk 4]
```diff
@@ -13,17 +13,20 @@ const assocNodes = nodes =>
...
-    const dirtyNodes = RecordReader.findNodesIndexed(Node.isDirty)(record)
+    const dirtyNodeUuids = NodesIndex.getNodeUuidsDirty(record)
 
     const nodesToUpdate = R.pipe(
       R.filter(
         n => {
-          const dirtyNode = R.prop(Node.getUuid(n), dirtyNodes)
+          const nodeUuid = Node.getUuid(n)
+          const dirtyNode = R.includes(nodeUuid, dirtyNodeUuids) && RecordReader.getNodeByUuid(nodeUuid)(record)
+
```
Comment (minotogna) [inline common/record/_internal/recordUpdater.js]: If is in the dirtyNodes index, it is also in the record or isn't it?

File: common/record/node.js [hunk 5]
```diff
@@ -27,12 +27,12 @@ const metaKeys = {
   hierarchy: 'h',
   childApplicability: 'childApplicability',
   defaultValue: 'defaultValue',
+  codeAttributeHierarchy: 'codeAttrH',
 }
```
Comment (minotogna) [inline common/record/node.js]: Should we call these keys hierarchy and hierachyCode?

File: common/record/_internal/recordNodesIndex.js (hunk 1 above)
Comment (fefrain) [reply common/record/_internal/recordNodesIndex.js]: right, it would be more clear

File: common/record/_internal/recordNodesIndex.js (hunk 2 above)
Comment (fefrain) [reply common/record/_internal/recordNodesIndex.js]: ok

**Pull request merged**